lib/mongoid/collection_configurable.rb

```diff
@@ -0,0 +1,30 @@
+# frozen_string_literal: true
+
+module Mongoid
+
+  # Encapsulates behavior around defining collections.
+  module CollectionConfigurable
+    extend ActiveSupport::Concern
+
+    module ClassMethods
+      def create_collection
+        if coll_options = collection.database.list_collections(filter: { name: collection_name.to_s }).first
+          logger.info(
+            "MONGOID: Collection '#{collection_name}' already exist " +
+            "in database '#{database_name}' with options '#{coll_options}'."
+          )
```

Comment:
I also think I would prefer this method to 1) retrieve collection options and 2) raise if they are different from current specification instead of succeeding in that case but leaving db in a state different from what one might expect. Thoughts?

Reply:
This would be ideal; however, implementing this adds some knowledge about the driver and collection options to Mongoid. When we create time series collection, we pass options using `time_series` key. When we get collection information using `listCollections`, the key is `timeseries`. Same with `expire_after` and `expireAfterSeconds`, `clustered_index` and `clusteredIndex`, and so on - https://github.com/mongodb/mongo-ruby-driver/blob/master/lib/mongo/collection.rb#L68

Further, if you do not specify, for example, `granularity` for a time series collection, it will be defaulted to seconds. However, what if a collection was created with with granularity = hours, but the corresponding Mongoid model omits the granularity option? Should we also validate defaults (=> adding even more knowledge about driver/server implementation details to the Mongoid)? Without such validation we cannot guarantee the state of db.

The above can be done in Mongoid, but we need to agree on it. I believe we usually avoid adding such low level details.

Reply:
I would propose that we should only check for conflicts in cases where values have been explicitly set by the user in Mongoid on the `store_in` macro, i.e. not validate vs Mongoid's defaults.

Reply:
@johnnyshields In the proposed implementation if we add validations to the ` db:mongoid:create_collections` task, it can be run as often as possible. It does not modify existing collections; so, it would be possible to execute it as a deployment step.

Reply:
I see, retrieving collection options and munching them is not something I think belongs in this PR.

I tested and collection cannot be created twice with identical options:
```
irb(main):006:0> a['foo1',capped:true,size:2].create
=> #<Mongo::Operation::Result:0x2660 documents=[{"ok"=>1.0}]>
irb(main):007:0> a['foo1',capped:true,size:2].create
/home/w/.rbenv/versions/3.1.0/lib/ruby/gems/3.1.0/gems/mongo-2.18.1/lib/mongo/operation/result.rb:364:in `raise_operation_failure': [48:NamespaceExists]: Collection already exists. NS: admin.foo1 (on localhost:14400) (Mongo::Error::OperationFailure)
	from /home/w/.rbenv/versions/3.1.0/lib/ruby/gems/3.1.0/gems/mongo-2.18.1/lib/mongo/operation/result.rb:330:in `validate!'
	from /home/w/.rbenv/versions/3.1.0/lib/ruby/gems/3.1.0/gems/mongo-2.18.1/lib/mongo/operation/shared/response_handling.rb:36:in `block (3 levels) in validate_result'
	from /home/w/.rbenv/versions/3.1.0/lib/ruby/gems/3.1.0/gems/mongo-2.18.1/lib/mongo/operation/shared/response_handling.rb:108:in `add_server_diagnostics'
	from /home/w/.rbenv/versions/3.1.0/lib/ruby/gems/3.1.0/gems/mongo-2.18.1/lib/mongo/operation/shared/response_handling.rb:35:in `block (2 levels) in validate_result'
	from /home/w/.rbenv/versions/3.1.0/lib/ruby/gems/3.1.0/gems/mongo-2.18.1/lib/mongo/operation/shared/response_handling.rb:54:in `add_error_labels'
	from /home/w/.rbenv/versions/3.1.0/lib/ruby/gems/3.1.0/gems/mongo-2.18.1/lib/mongo/operation/shared/response_handling.rb:34:in `block in validate_result'
	from /home/w/.rbenv/versions/3.1.0/lib/ruby/gems/3.1.0/gems/mongo-2.18.1/lib/mongo/operation/shared/response_handling.rb:94:in `unpin_maybe'
	from /home/w/.rbenv/versions/3.1.0/lib/ruby/gems/3.1.0/gems/mongo-2.18.1/lib/mongo/operation/shared/response_handling.rb:33:in `validate_result'
	from /home/w/.rbenv/versions/3.1.0/lib/ruby/gems/3.1.0/gems/mongo-2.18.1/lib/mongo/operation/shared/executable.rb:81:in `block in execute'
	from <internal:kernel>:90:in `tap'
	from /home/w/.rbenv/versions/3.1.0/lib/ruby/gems/3.1.0/gems/mongo-2.18.1/lib/mongo/operation/shared/executable.rb:80:in `execute'
	from /home/w/.rbenv/versions/3.1.0/lib/ruby/gems/3.1.0/gems/mongo-2.18.1/lib/mongo/operation/shared/polymorphic_operation.rb:50:in `execute_with_connection'
	from /home/w/.rbenv/versions/3.1.0/lib/ruby/gems/3.1.0/gems/mongo-2.18.1/lib/mongo/operation/shared/polymorphic_operation.rb:36:in `block in execute'
	from /home/w/.rbenv/versions/3.1.0/lib/ruby/gems/3.1.0/gems/mongo-2.18.1/lib/mongo/server/connection_pool.rb:667:in `with_connection'
	from /home/w/.rbenv/versions/3.1.0/lib/ruby/gems/3.1.0/gems/mongo-2.18.1/lib/mongo/server.rb:448:in `with_connection'
	from /home/w/.rbenv/versions/3.1.0/lib/ruby/gems/3.1.0/gems/mongo-2.18.1/lib/mongo/operation/shared/polymorphic_operation.rb:35:in `execute'
	... 9 levels...
irb(main):008:0> 
```

How about:

`Model.create_collection` creates the collection if it does not exist
`Model.create_collection!` attempts to create the collection every time, will fail if collection exists

+ have two rake tasks, maybe `mongoid:create_collections` and `mongoid:create_collections:force` or something like that.

@Neilshweky

Reply:
Should bear in mind users are lazy creatures, and they are not going to want to run a Rake task to create collections. Today everyone is used having it happen automatically when the first model instance is persisted at runtime. The Rake task is useful as a diagnostic tool perhaps, but should not be the golden path.

Perhaps the first time a given collection is accessed in a process, Mongoid could soft-validate that the collection's options match what is actually in the DB, and output a warning log line saying "either revert the options or make a new collection" (AFAIK collections can't be reconfigured.)

Reply:
`mongoid:create_collections:force` in the Rails sense would be useful for development. The `mongoid:create_collections:validate` I think is impractical - it would always fail in production and in development on a non-empty database, for the empty database in development `mongoid:create_collections:force` would work just as well. The `mongoid:create_collections` with warning behavior I think could be useful in production and would be the equivalent to Rails' migrations, alternatively it could perhaps be renamed to `mongoid:create_collections:missing` and not warn about the existing ones.

For indexes we have precedent of checking indexes in db vs indexes declared in models, this is also non-trivial and from what I remember doesn't quite work 100% of the time.